elliptic-curve/src/arithmetic.rs

```diff
@@ -84,3 +87,28 @@ pub trait PrimeCurveArithmetic:
     /// Prime order elliptic curve group.
     type CurveGroup: group::prime::PrimeCurve<Affine = <Self as CurveArithmetic>::AffinePoint>;
 }
+
+/// Normalize point(s) in projective representation by converting them to their affine ones.
+pub trait Normalize: group::Curve {
+    /// Perform a batched conversion to affine representation on a sequence of projective points
+    /// at an amortized cost that should be practically as efficient as a single conversion.
+    /// Internally, implementors should rely upon `InvertBatch`.
+    /// This variation takes a const-generic array and thus does not require `alloc`.
+    fn batch_normalize_array<const N: usize>(points: &[Self; N]) -> [Self::AffineRepr; N];
+
+    /// Perform a batched conversion to affine representation on a sequence of projective points
+    /// at an amortized cost that should be practically as efficient as a single conversion.
+    /// Internally, implementors should rely upon `InvertBatch`.
+    /// This variation takes a (possibly dynamically allocated) slice and returns `FromIterator<Self::AffinePoint>`
+    /// allowing it to work with any container.
+    /// However, this also requires to make dynamic allocations and as such requires `alloc`.
+    #[cfg(feature = "alloc")]
+    fn batch_normalize_to_vec(points: &[Self]) -> Vec<Self::AffineRepr>
+    where
+        Self::AffineRepr: Copy + Default,
+    {
+        let mut ret = vec![Self::AffineRepr::default(); points.len()];
+        Self::batch_normalize(points, &mut ret);
+        ret
+    }
+}
```

Comment:
Perhaps it's possible to use a generic trait to abstract over the difference between these two, e.g. something generic over the input or output types:

```rust
pub trait Normalize<Points>: group::Curve {
    type Output: AsRef<Self::AffineRepr>;

    // TODO: name that doesn't clash with `group::Curve::batch_normalize`
    fn batch_normalize(&Points) -> Self::Output;
}
```

Then you could have something like:

```rust
#[cfg(feature = "alloc")]
impl<Point: group::Curve> Normalize<[Point]> for Point {
    type Output = Vec<Point::AffineRepr>;

    fn batch_normalize(&[Point]) -> Vec<Point::AffineRepr> {
        let mut ret = vec![Self::AffineRepr::default(); points.len()];
        Self::batch_normalize(points, &mut ret);
        ret
    }
} 
```

Reply:
I love this, I refactored. I don't agree that we should have a different name than `group::Curve::batch_normalize`; it happens that there are colliding names for different names, this is the appropriate name. Users can choose which API they wish to use, and the implementation is going to default to the same code anyways. In any case, I'll rename if you wish, but I really need to wrap this up soon as it became much more work than originally intended and I have other priorities. Thank you